File: lib/soonex/mode.ex
defmodule Soonex.Mode do
  @moduledoc false

  def head_script do
    Phoenix.HTML.raw("""
    <script>
      try {
        const m = localStorage.getItem("data-mode");
        const prefersDark = matchMedia("(prefers-color-scheme: dark)").matches;
        const mode = m === "dark" || m === "light" ? m : (prefersDark ? "dark" : "light");
        document.documentElement.setAttribute("data-mode", mode);
      } catch (_) {}
    </script>
    """)
  end

  def current(assigns) do
    case Map.get(assigns, :mode) do
      "dark" -> "dark"
      _ -> "light"
    end
  end
end

File: assets/js/mode.js
import { bindStoredPreference } from "./controls-shared.js"

const systemMode = () =>
  matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light"

const resolveMode = (raw) =>
  raw === "dark" || raw === "light" ? raw : systemMode()

bindStoredPreference({
  attr: "data-mode",
  controlId: "mode-switcher",
  resolve: resolveMode,
  syncControl: (mode) => {
    const root = document.getElementById("mode-switcher")
    if (!root) return
    const pressed = mode === "dark"
    const button = root.querySelector('[data-scope="toggle"][data-part="root"]')
    if (button && (button.getAttribute("data-state") === "on") === pressed) return
    root.dispatchEvent(
      new CustomEvent("corex:toggle:set-pressed", {
        bubbles: false,
        detail: { pressed },
      }),
    )
  },
  clientEvent: "corex:set-mode",
  parseClientEvent: (event) => {
    const { pressed, value } = event.detail ?? {}
    if (typeof pressed === "boolean") return pressed ? "dark" : "light"
    return Array.isArray(value) && value.includes("dark") ? "dark" : "light"
  },
})
